File: clients/csharp/Model/Obstacle.cs
namespace AiCup22.Model
{
    /// <summary>
    /// An obstacle on the map
    /// </summary>
    public struct Obstacle
    {
        /// <summary>
        /// Unique id
        /// </summary>
        public int Id { get; set; }
        /// <summary>
        /// Center position
        /// </summary>
        public AiCup22.Model.Vec2 Position { get; set; }
        /// <summary>
        /// Obstacle's radius
        /// </summary>
        public double Radius { get; set; }
        /// <summary>
        /// Whether units can see through this obstacle, or it blocks the view
        /// </summary>
        public bool CanSeeThrough { get; set; }
        /// <summary>
        /// Whether projectiles can go through this obstacle
        /// </summary>
        public bool CanShootThrough { get; set; }
    
        public Obstacle(int id, AiCup22.Model.Vec2 position, double radius, bool canSeeThrough, bool canShootThrough)
        {
            this.Id = id;
            this.Position = position;
            this.Radius = radius;
            this.CanSeeThrough = canSeeThrough;
            this.CanShootThrough = canShootThrough;
        }
    
        /// <summary> Read Obstacle from reader </summary>
        public static Obstacle ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Obstacle();
            result.Id = reader.ReadInt32();
            result.Position = AiCup22.Model.Vec2.ReadFrom(reader);
            result.Radius = reader.ReadDouble();
            result.CanSeeThrough = reader.ReadBoolean();
            result.CanShootThrough = reader.ReadBoolean();
            return result;
        }
    
        /// <summary> Write Obstacle to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(Id);
            Position.WriteTo(writer);
            writer.Write(Radius);
            writer.Write(CanSeeThrough);
            writer.Write(CanShootThrough);
        }
    
        /// <summary> Get string representation of Obstacle </summary>
        public override string ToString() {
            string stringResult = "Obstacle { ";
            stringResult += "Id: ";
            stringResult += Id.ToString();
            stringResult += ", ";
            stringResult += "Position: ";
            stringResult += Position.ToString();
            stringResult += ", ";
            stringResult += "Radius: ";
            stringResult += Radius.ToString();
            stringResult += ", ";
            stringResult += "CanSeeThrough: ";
            stringResult += CanSeeThrough.ToString();
            stringResult += ", ";
            stringResult += "CanShootThrough: ";
            stringResult += CanShootThrough.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}

File: clients/fsharp/Model/Obstacle.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Model

open AiCup22

/// An obstacle on the map
type Obstacle = {
    /// Unique id
    Id: int;
    /// Center position
    Position: Model.Vec2;
    /// Obstacle's radius
    Radius: double;
    /// Whether units can see through this obstacle, or it blocks the view
    CanSeeThrough: bool;
    /// Whether projectiles can go through this obstacle
    CanShootThrough: bool;
} with

    /// Write Obstacle to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.Id
        this.Position.writeTo writer
        writer.Write this.Radius
        writer.Write this.CanSeeThrough
        writer.Write this.CanShootThrough
        ()

    /// Read Obstacle from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Id = reader.ReadInt32()
        Position = Model.Vec2.readFrom reader;
        Radius = reader.ReadDouble()
        CanSeeThrough = reader.ReadBoolean()
        CanShootThrough = reader.ReadBoolean()
    }